1.14 Fluid rework.

gigaherz opened the pull request:
May or may not be what forge wants.

gigaherz committed:
Fluid rework. First pass. Very WIP.

gigaherz committed:
Oops, for some reason the patches didn't generate.
Readd the FluidBucketWrapper, that's still needed with tweaks.

KingLemming commented:
Well, I had been working on this offline but lemme see if I can't reconcile this with what I was thinking.

gigaherz committed:
Rudimentary logic to allow rendering custom fluids with the vanilla fluid rendering code. Fluid models are not supported yet.

gigaherz committed:
Fix the color value in the vanilla attribute instances. Fix the test fluid not recognizing itself.

gigaherz committed:
Small tweak to the test mod.

mcenderdragon commented:
Looks good, does the fluid correctly render and flow ingame?

gigaherz commented:
Yes, but with the caveat only the vanilla renderer is supported, no custom forge fluid model. I have been waiting for KL to see how things move forward.

KingLemming committed:
Sanity check before storm.

Signed-off-by: King Lemming <kinglemming@gmail.com>

KingLemming committed:
Initial changes to FluidAction.

Signed-off-by: King Lemming <kinglemming@gmail.com>

KingLemming committed:
Lots of updates and cleanups.

FluidUtil class is borked - we *probably* need to think about what is truly required there.

Signed-off-by: King Lemming <kinglemming@gmail.com>

gigaherz committed:
Fix some build errors after the merge.

ichttt commented on src/main/java/net/minecraftforge/client/ForgeClientHandler.java:
```diff
@@ -24,27 +24,7 @@
 import net.minecraftforge.eventbus.api.SubscribeEvent;
-import net.minecraftforge.fluids.FluidRegistry;
 
 public class ForgeClientHandler
```
This class is dead now, might as well kill it

ichttt commented on patches/minecraft/net/minecraft/fluid/LavaFluid.java.patch:
```diff
@@ -0,0 +1,11 @@
+       return 100.0F;
+    }
+ 
++   public net.minecraftforge.fluids.FluidAttributes createAttributes(Fluid fluid) { return net.minecraftforge.fluids.FluidAttributes.VANILLA_LAVA; }
```
Should stick an `@Override` before this, just to make sure

gigaherz replied:
Good catch.

ichttt commented on patches/minecraft/net/minecraft/fluid/WaterFluid.java.patch:
```diff
@@ -0,0 +1,11 @@
+       return p_215665_5_ == Direction.DOWN && !p_215665_4_.func_207185_a(FluidTags.field_206959_a);
+    }
+ 
++   public net.minecraftforge.fluids.FluidAttributes createAttributes(Fluid fluid) { return net.minecraftforge.fluids.FluidAttributes.VANILLA_WATER; }
```
Same here, I'd use `@Override`

ichttt commented on src/main/java/net/minecraftforge/fluids/IFluidTank.java:
```diff
@@ -1,33 +1,16 @@
-/*
```
Why was this removed?

gigaherz replied:
I believe KL copypasted the entire file from his draft implementation elsewhere, and that draft didn't have the license headers in it.

ichttt commented on src/main/java/net/minecraftforge/fluids/capability/templates/FluidTank.java:
```diff
@@ -0,0 +1,229 @@
+/*
```
Forge License Header please?

gigaherz commented on src/main/java/net/minecraftforge/fluids/FluidAttributes.java:
```diff
@@ -54,17 +56,31 @@
+            .setDensity(0).setTemperature(0).setLuminosity(0);
+    public static final FluidAttributes VANILLA_WATER = new FluidAttributes("water", LOCATION_WATER_STILL, LOCATION_WATER_FLOW, LOCATION_WATER_OVERLAY, -1);
+    public static final FluidAttributes VANILLA_LAVA = new FluidAttributes("lava", LOCATION_LAVA_STILL, LOCATION_LAVA_FLOW, -1)
+            /* TODO: set temperature and stuff */;
```
Note to self: fix this before the PR gets merged.

ichttt reviewed (commented):
Just some small things I found while reading, not a full review

gigaherz commented on src/main/java/net/minecraftforge/fluids/IFluidTank.java:
```diff
@@ -1,33 +1,16 @@
- * License along with this library; if not, write to the Free Software
- * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
- */
-
```
This probably needs to be restored?

gigaherz commented on src/main/java/net/minecraftforge/fml/ModLoader.java:
```diff
@@ -212,7 +212,7 @@ private void accumulateErrors(List<ModLoadingException> errors) {
             LOGGER.fatal(LOADING,"File {} constructed {} mods: {}, but had {} mods specified: {}",
                     modFile.getFilePath(),
-                    containers.size(), containers.stream().map(ModContainer::getModId).collect(Collectors.toList()),
+                    containers.size(), containers.stream().map(c -> c != null ? c.getModId() : "(null)").collect(Collectors.toList()),
```
@cpw I'm assuming this is not meant to contain nulls here. I made this change here to allow a completely opaque crash to become a meaninful crash, in case a Mod annotation does not have a matching entry in the mods.toml file.

gigaherz reviewed (commented):
I can't "request changes" on my own PR, but these are things I think need to be looked at before we merge.

gigaherz committed:
Run the updateLicenseCustomMain task to add license headers everywhere that matters.
Fix the attributes for lava.
Implement an alternative to the block field in FluidAttributes.

gigaherz committed:
Move createAttributes to the Fluid class and make it protected, so that it can't be called from outside.
Make getAttributes final in Fluid, so that it can't be overriden by mistake.

gigaherz committed:
Move FluidAttributes constructing to a builder. Remove some patches that I should have removed in the previous commit.

pupnewfster commented on src/main/java/net/minecraftforge/fluids/FluidStack.java:
```diff
@@ -123,14 +124,14 @@ public final Fluid getFluid()
     }
 
-    public String getLocalizedName()
+    public String getDisplayName()
```
Would it make more sense to have this be an `ITextComponent` like `Item#getDisplayName(ItemStack stack)` is?

gigaherz replied:
Done. Good idea.

TeamSpen210 commented on patches/minecraft/net/minecraft/fluid/Fluid.java.patch:
```diff
@@ -18,5 +26,36 @@
++   protected net.minecraftforge.fluids.FluidAttributes createAttributes(Fluid fluid)
++   {
++      if (fluid instanceof EmptyFluid)
++         return net.minecraftforge.fluids.FluidAttributes.builder("water", null, null)
```
Shouldn't this ID be `"empty"`, not `"water"`?
```suggestion
+         return net.minecraftforge.fluids.FluidAttributes.builder("empty", null, null)
```

gigaherz replied:
Yes. Nice catch.

gigaherz committed:
Change getDisplayName to return ITextComponent. Fix copypaste mistake.

LexManos commented:
This PR, and it's sister [Recipe Rewrite](https://github.com/MinecraftForge/MinecraftForge/pull/6014) are both good to go IMO. And can be merged, I just want them merged at the same time/quickly together. As they are both major binary breaking changes.

Once these two are in, we'll do a sweep of basic bugs/prs and call it good for a RB.

MajorTuvok commented on src/main/java/net/minecraftforge/fluids/capability/templates/EmptyFluidHandler.java:
```diff
@@ -19,69 +19,47 @@
-    {
-        return 0;
-    }
+    public FluidStack getFluidInTank(int tank) { return null; }
```
What about a non-null policy similiar to ItemStack, with an isEmpty() chexk to find out, whether the FluidStack is empty?

gigaherz commented:
Merged in b7e1cc5f6

oOMitchOo commented:
Since issue #5153 is referenced here I would like to know, how this makes Forge added fluids placeable into 'waterloggable' blocks? It's my understanding that IWaterLoggable is hardcoded to work with Minecraft water, and since it translates to a boolean property (waterlogged) in the blocks, which can get waterlogged, there is not really a way to make your own fluids placeable into 'waterloggable' blocks right now?

gigaherz commented:
That is correct. Waterlogged blocks are hardcoded to 1 fluid. In the case of vanilla blocks, they are hardcoded to water, so they won't accept your fluid. You could make new blocks that waterlog to other things, but in any case, you will only ever have a finite set of fluids, because the fluid system requires the property values be known ahead of time.
Unnoen made a mistake there. The issue should NOT have been closed.
